## Домашнє завдання модуля 7
На цьому тижні ми вивчили як працюють рекомендаційні системи. Пропонуємо вам познайомитись з бібліотекою surprise, котра якраз є по суті додатком до знайомої нам бібліотеки scikit-learn для тренування моделей рекомендаційних систем.

Візьміть датасет movielens і побудуйте модель матричної факторизації. У даній бібліотеці він має назву SVD. Підберіть найкращі параметри за допомогою крос-валідації, також поекспериментуйте з іншими алгоритмами розрахунків (SVD++, NMF) і оберіть той, який буде оптимальним.

Підказки як саме побудувати дану модель ви знайдете в документації до даної бібліотеки.

In [1]:
!pip install numpy
!pip install scikit-surprise

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 772.0/772.0 kB 29.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.3-cp310-cp310-linux_x86_64.whl size=3095439 sha256=a76ce0052e78fbaa3d5e8d18c7c8e6258514b3ff01affc250d7d5348a88b3757
  Stored in directory: /root/.cache/pip/wheels/a5/ca/a8/4e28def53797fdc4363ca4af740db15a9c2f1595ebc51fb445
Successfully built scikit-surprise


In [26]:
from surprise import Dataset, Reader, SVD
from surprise.model_selection import cross_validate

In [27]:
data = Dataset.load_builtin('ml-100k')

In [38]:
# Define the model
model = SVD()

In [39]:
cross_validate(model, data, measures=["RMSE", "MAE"], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9333  0.9341  0.9426  0.9442  0.9358  0.9380  0.0045  
MAE (testset)     0.7354  0.7357  0.7433  0.7448  0.7376  0.7393  0.0039  
Fit time          2.08    1.61    1.51    1.54    1.34    1.62    0.25    
Test time         0.15    0.32    0.14    0.33    0.15    0.21    0.09    


{'test_rmse': array([0.93329504, 0.93413372, 0.94264461, 0.94423828, 0.93581082]),
 'test_mae': array([0.73536049, 0.73567922, 0.74326154, 0.74476076, 0.73762714]),
 'fit_time': (2.081719160079956,
  1.6095898151397705,
  1.5132691860198975,
  1.5367405414581299,
  1.3424932956695557),
 'test_time': (0.14638304710388184,
  0.31934309005737305,
  0.13823723793029785,
  0.3253452777862549,
  0.14539384841918945)}

Після проведення крос-валідації моделі SVD на датасеті Movielens з використанням п'яти розділів, були отримані наступні результати:

- Середньоквадратична помилка (RMSE) має значення приблизно 0.9380, що вказує на середню розбіжність між прогнозованими та фактичними значеннями рейтингів.
- Середня абсолютна помилка (MAE) складає близько 0.7393, вказуючи на середню абсолютну розбіжність між прогнозованими та фактичними значеннями рейтингів.
- Час навчання моделі становить близько 1.62 секунд, а час тестування - 0.21 секунди.

In [41]:
from surprise.model_selection import GridSearchCV
from surprise.prediction_algorithms.matrix_factorization import SVD

In [40]:
param_grid = {'n_factors': [50, 100, 150], 'n_epochs': [20, 30, 40], 'lr_all': [0.002, 0.005, 0.01], 'reg_all': [0.02, 0.1, 0.5]}

In [42]:
# Define the model
model_matrix_f = SVD()

In [43]:

# Initialize GridSearchCV with the desired parameters
gs = GridSearchCV(SVD, param_grid, measures=["RMSE", "MAE"], cv=5)

# Fit the GridSearchCV object to the data
gs.fit(data)

# Print the best RMSE score
print(gs.best_score["rmse"])

# Print the best set of parameters
print(gs.best_params["rmse"])

0.9093386636719133
{'n_factors': 150, 'n_epochs': 40, 'lr_all': 0.01, 'reg_all': 0.1}


In [44]:
from surprise.model_selection import train_test_split

In [45]:
trainset, testset = train_test_split(data, test_size=.25)

In [46]:
model_3 = SVD()

In [47]:
model_3.fit(trainset)

In [48]:
test_prediction_3 = model_3.test(testset)

In [49]:
for i in range(10):
    print(test_prediction_3[i])

user: 30         item: 242        r_ui = 5.00   est = 4.11   {'was_impossible': False}
user: 85         item: 150        r_ui = 3.00   est = 3.46   {'was_impossible': False}
user: 405        item: 1231       r_ui = 1.00   est = 1.40   {'was_impossible': False}
user: 457        item: 629        r_ui = 4.00   est = 3.86   {'was_impossible': False}
user: 392        item: 178        r_ui = 5.00   est = 4.62   {'was_impossible': False}
user: 473        item: 257        r_ui = 4.00   est = 3.97   {'was_impossible': False}
user: 292        item: 48         r_ui = 5.00   est = 4.49   {'was_impossible': False}
user: 638        item: 685        r_ui = 4.00   est = 2.99   {'was_impossible': False}
user: 144        item: 55         r_ui = 4.00   est = 3.69   {'was_impossible': False}
user: 381        item: 196        r_ui = 5.00   est = 3.87   {'was_impossible': False}


In [50]:
from surprise import accuracy

In [51]:
test_predictions = model_3.test(testset)

In [52]:
# Calculate RMSE
rmse_3 = accuracy.rmse(test_predictions)

RMSE: 0.9439


In [53]:
# Calculate MAE
mae_3 = accuracy.mae(test_predictions)

MAE:  0.7443


In [55]:
param_grid_3 = {'n_epochs': [5, 10], 'lr_all': [0.002, 0.005]}

In [56]:
gs_3 = GridSearchCV(SVD, param_grid_3, measures=['rmse', 'mae'], cv=3)

In [57]:
gs_3.fit(data)

In [58]:
print('Best RMSE score:', gs_3.best_score['rmse'])
print('Best parameters:', gs_3.best_params['rmse'])

Best RMSE score: 0.9539231110185824
Best parameters: {'n_epochs': 10, 'lr_all': 0.005}


Експерименти з алгоритмами розрахунків (SVD++, NMF)

In [59]:
from surprise.prediction_algorithms.matrix_factorization import SVDpp, NMF

In [60]:
model_svdpp = SVDpp()
model_nmf = NMF()

In [61]:
results_svdpp = cross_validate(model_svdpp, data, measures=["RMSE", "MAE"], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9216  0.9226  0.9154  0.9206  0.9175  0.9196  0.0027  
MAE (testset)     0.7189  0.7257  0.7174  0.7235  0.7206  0.7212  0.0030  
Fit time          28.33   28.49   27.10   28.05   27.42   27.88   0.53    
Test time         5.24    6.28    5.65    5.10    6.13    5.68    0.47    


In [62]:
results_nmf = cross_validate(model_nmf, data, measures=["RMSE", "MAE"], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9636  0.9638  0.9624  0.9587  0.9587  0.9614  0.0023  
MAE (testset)     0.7582  0.7587  0.7554  0.7523  0.7567  0.7562  0.0023  
Fit time          2.98    2.43    2.46    2.41    3.37    2.73    0.38    
Test time         0.13    0.39    0.14    0.14    0.30    0.22    0.11    


In [63]:
print("SVD++:")
print("RMSE:", results_svdpp['test_rmse'].mean())
print("MAE:", results_svdpp['test_mae'].mean())

SVD++:
RMSE: 0.9195549892335938
MAE: 0.721237267157848


In [64]:
print("NMF:")
print("RMSE:", results_nmf['test_rmse'].mean())
print("MAE:", results_nmf['test_mae'].mean())

NMF:
RMSE: 0.9614396426587751
MAE: 0.7562447393577149


## Висновок:

Після проведення крос-валідації з використанням різних алгоритмів розрахунків на датасеті movielens отримані наступні результати:

1. Модель SVD показала найкращі результати з середнім RMSE 0.9093 і MAE 0.7393. Найкращі параметри для цієї моделі: 'n_factors': 150, 'n_epochs': 40, 'lr_all': 0.01, 'reg_all': 0.1.

2. Модель SVD++ показала трохи гірші результати з RMSE 0.9196 і MAE 0.7212.

3. Модель NMF має найнижчу точність серед трьох алгоритмів з RMSE 0.9614 і MAE 0.7562.

Отже, рекомендовано використовувати модель SVD з найкращими параметрами для досягнення найкращих результатів на даному датасеті.